compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -978,6 +1019,22 @@ impl<'tcx> InferCtxt<'tcx> {
         }
     }
 
+    pub fn push_hir_typeck_potentially_region_dependent_goal(
+        &self,
+        goal: PredicateObligation<'tcx>,
+    ) {
+        let mut inner = self.inner.borrow_mut();
+        inner.undo_log.push(UndoLog::PushHirTypeckPotentiallyRegionDependentGoal);
+        inner.hir_typeck_potentially_region_dependent_goals.push(goal);
```

Comment:
questions:
We only let people evaluate goals inside of a probe if the fulfillmentctxt/whatever was also created in that probe? 
This codepath just exists to not (incorrectly) leak details from that previous scenario, to outside of the probe? 
We don't actually have support for "rolling back" trait solving?

Reply:
we do when using a separate fulfillment context inside of a probe